elements/pl-matching/pl-matching.py

```diff
@@ -0,0 +1,440 @@
+    """Converts a base-10 integer to base 26, using letter_a as the "0" digit."""
+    result = ''
+    while i > 0:
+        i -= 1
+        result = chr(ord(letter_a) + i % 26) + result
+        i //= 26
+
+    return result
+
+
+def get_counter(i, counter_type):
+    """Converts an integer counter to the specified CSS counter type"""
+    if counter_type == 'lower-alpha':
+        return alpha_counter(i, 'a')
+    elif counter_type == 'upper-alpha':
+        return alpha_counter(i, 'A')
+    elif counter_type == 'decimal':
+        return str(i)
+    else:
+        raise Exception('Illegal counter-type in pl-matching element: "' + counter_type + '" should be "decimal", "lower-alpha" or "upper-alpha".')
+
+
+def legal_answer(answer, counter_type, options):
+    """Checks that the given answer is within the range of the given counter type."""
+    if counter_type == 'lower-alpha':
+        return 'a' <= answer <= chr(ord('a') + len(options))
+    elif counter_type == 'upper-alpha':
+        return 'A' <= answer <= chr(ord('A') + len(options))
+    elif counter_type == 'decimal':
+        return 1 <= answer <= len(options)
+    else:
+        return False
+
+
+def get_select_options(options_list, selected_value):
+    return [{
+        'value': opt,
+        'selected': 'selected' if opt == selected_value else ''
+    } for opt in options_list]
+
+
+def partition(data, pred):
+    """Implements a partition function, splitting the data into two lists based on the predicate."""
+    yes, no = [], []
+    for d in data:
+        if pred(d):
+            yes.append(d)
+        else:
+            no.append(d)
+    return (yes, no)
+
+
+def categorize_matches(element, data):
+    """Get provided statements and options from the pl-matching element"""
+    options = {}
+    statements = []
+    index = 0
+
+    # Sort the elements so that pl-options come first.
+    children = element[:]
+    children.sort(key=lambda child: child.tag)
+
+    for child in children:
+        if child.tag in ['pl-option', 'pl_option']:
+            pl.check_attribs(child, required_attribs=[], optional_attribs=['name'])
+            child_html = pl.inner_html(child)
+            option_name = pl.get_string_attrib(child, 'name', child_html)
+
+            # An option object has: index of appearance in the pl-matching element;
+            # the name attribute; and the html content.
+            option = {'index': index, 'name': option_name, 'html': child_html}
+            options[option_name] = option
+            index += 1
+
+        elif child.tag in ['pl-statement', 'pl_statement']:
+            pl.check_attribs(child, required_attribs=['match'], optional_attribs=[])
+            child_html = pl.inner_html(child)
+            match_name = pl.get_string_attrib(child, 'match')
+            if match_name not in options:
+                raise Exception(f'pl-statement "match" attribute of {match_name} does not match any of the pl-option elements.')
```

Comment:
Not saying this must be done, just raising it as an option: what if a `pl-option` could be implied if it doesn't exist? In that case, `pl-option` would be optional, only used for (1) names that link to longer html, or (2) distractors. This could make the most common cases (where the option is a simple string) easier to create. In your example question, options like `Paris` and `Mexico City` wouldn't need to be added, but options like `d.c.` would be added to change the html of the option, and `New York City` would be a distractor. Thoughts?

Reply:
That's a lovely idea.